Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php

```diff
@@ -0,0 +1,171 @@
+                        NodeVariantSelectionStrategy::STRATEGY_ALL_SPECIALIZATIONS
+                    )
+                );
+
+                $result = HandlingResult::createWithEnabled($node);
+                $handledNodeResults->add($result);
+                $this->logResult($result);
+
+            } elseif ($this->needsDisabling($node, $now) && !$this->isHidden($node, $nodeHiddenStateFinder)) {
+                $contentRepository->handle(
+                    DisableNodeAggregate::create(
+                        $node->subgraphIdentity->contentStreamId,
+                        $node->nodeAggregateId,
+                        $node->subgraphIdentity->dimensionSpacePoint,
+                        NodeVariantSelectionStrategy::STRATEGY_ALL_SPECIALIZATIONS
+                    )
+                );
+
+                $result = HandlingResult::createWithDisabled($node);
+                $handledNodeResults->add($result);
+                $this->logResult($result);
+            }
+        }
+        return $handledNodeResults;
+    }
+
+    /**
+     * @return \Generator<Node>
+     */
+    private function getNodesWithExceededDates(ContentRepository $contentRepository, Workspace $liveWorkspace, \DateTimeImmutable $now): \Generator
+    {
+        $dimensionSpacePoints = $contentRepository->getVariationGraph()->getDimensionSpacePoints();
+
+        foreach ($dimensionSpacePoints as $dimensionSpacePoint) {
+
+            $contentGraph = $contentRepository->getContentGraph();
+
+            // We fetch without restriction to get also all disabled nodes
+            $subgraph = $contentGraph->getSubgraph(
+                $liveWorkspace->currentContentStreamId,
+                $dimensionSpacePoint,
+                VisibilityConstraints::withoutRestrictions()
+            );
+
+            $rootNode = $subgraph->findRootNodeByType(NodeTypeName::fromString('Neos.Neos:Sites'));
+
+            $nodes = $subgraph->findDescendantNodes(
+                $rootNode->nodeAggregateId,
+                FindDescendantNodesFilter::create(
+                    nodeTypes: NodeTypeCriteria::createWithAllowedNodeTypeNames(NodeTypeNames::fromStringArray(['Neos.TimeableNodeVisibility:Timeable'])),
+                    propertyValue: OrCriteria::create(
+                        PropertyValueLessThanOrEqual::create(PropertyName::fromString('enableAfterDateTime'), $now->format(\DateTime::RFC3339)),
+                        PropertyValueLessThanOrEqual::create(PropertyName::fromString('disableAfterDateTime'), $now->format(\DateTime::RFC3339)),
+                    )
+
+                )
+            );
+
+            foreach ($nodes as $node) {
+
+                if (!$node->originDimensionSpacePoint->equals($dimensionSpacePoint)) {
+                    // The node will be enabled by node in origin dimension spacepoint
+                    continue;
+                }
+
+                yield $node;
+            }
```

Comment:
hmm that seems a bit costy to iterate over all nodes in all dimensions, and ditching a lot of nodes again, just to get the origin nodes?

I dont know how but there might be a better way? ^^ // cc @nezaniel

Reply:
Yes, the whole fetching is very expensive. But AFAIK we have not other option here.

Reply:
I think we hsould have a way to filter for origin nodes only, on a technical level that definitely makes sense.

Reply:
As we currently have no better way and this is just executes once per minute/5 minute, I guess this is fine and can get improved later. WDYT @kitsunet @mhsdesign ?

Reply:
Yes sure. Lets create a followup issue where we collect this and other issues (and please link it in the code :D)
Then we wont block this pr any further?